core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java

```diff
@@ -218,22 +249,23 @@ public ThreadPool(Settings settings) {
         int halfProcMaxAt5 = Math.min(((availableProcessors + 1) / 2), 5);
         int halfProcMaxAt10 = Math.min(((availableProcessors + 1) / 2), 10);
         Map<String, Settings> defaultExecutorTypeSettings = new HashMap<>();
-        add(defaultExecutorTypeSettings, new ExecutorSettingsBuilder(Names.GENERIC).size(4 * availableProcessors).keepAlive("30s"));
-        add(defaultExecutorTypeSettings, new ExecutorSettingsBuilder(Names.INDEX).size(availableProcessors).queueSize(200));
-        add(defaultExecutorTypeSettings, new ExecutorSettingsBuilder(Names.BULK).size(availableProcessors).queueSize(50));
-        add(defaultExecutorTypeSettings, new ExecutorSettingsBuilder(Names.GET).size(availableProcessors).queueSize(1000));
-        add(defaultExecutorTypeSettings, new ExecutorSettingsBuilder(Names.SEARCH).size(((availableProcessors * 3) / 2) + 1).queueSize(1000));
-        add(defaultExecutorTypeSettings, new ExecutorSettingsBuilder(Names.MANAGEMENT).size(5).keepAlive("5m"));
+        int genericThreadPoolMax = boundedBy(4 * availableProcessors, 128, 512);
+        add(defaultExecutorTypeSettings, new ScalingExecutorSettingsBuilder(Names.GENERIC).min(4).max(genericThreadPoolMax).keepAlive("30s"));
```

Comment:
@bleskes @ywelsch Always keeping some threads alive will help minimize latency from thread creation.

Each cached thread pool has a core pool size (the `min`) and a maximum pool size. When fewer than `min` worker threads are running, the JVM will create new threads. When `min` worker threads are running, the JVM prefers to queue requests rather than creating new worker threads. The JVM will only resort to creating creating new worker threads above `min` if the queue saturates.

This also explains why we use the [`ExecutorScalingQueue`](https://github.com/elastic/elasticsearch/pull/17017#discussion_r60527516). We first try to transfer to an active thread. If there is not one, we see if the pool has excess capacity. If it does, then we fake that the queue rejects the task to force the JVM to create a new worker thread rather than queuing the task. We only queue if there is no excess capacity in the thread pool.